### PR TITLE
WIP: VTOL transitioning

### PR DESCRIPTION
**Pull request by tumbili**
During a transition of the VTOL the attitude control module should publish the desired attitude setpoint, not any of the fixed wing or multicopter modules. A very simple transition strategy has only been implemented for the tiltrotor now.
The diff looks big but that's because the vtol_cleanup branch has not been merged into master yet.
This will need heavy bench testing first!!!

### CODE REVIEW

File: src/modules/vtol_att_control/tailsitter.cpp [hunk 1]
```diff
@@ -152,38 +140,41 @@ Tailsitter::scale_mc_output()
...
+
+			_actuators_out_1->control[actuator_controls_s::INDEX_ROLL] = -_actuators_fw_in->control[actuator_controls_s::INDEX_ROLL];	// roll elevon
+			_actuators_out_1->control[actuator_controls_s::INDEX_PITCH] = _actuators_fw_in->control[actuator_controls_s::INDEX_PITCH] + _params->fw_pitch_trim;	// pitch elevon
+			_actuators_out_1->control[actuator_controls_s::INDEX_YAW] = _actuators_fw_in->control[actuator_controls_s::INDEX_YAW];	// yaw
+			_actuators_out_1->control[actuator_controls_s::INDEX_THROTTLE] = _actuators_fw_in->control[actuator_controls_s::INDEX_THROTTLE];	// throttle
+			break;
+		case TRANSITION:
+		case EXTERNAL:
+			// not yet implemented, we are switching brute force at the moment
+			break;
```
Comment (bansiesta) [inline src/modules/vtol_att_control/tailsitter.cpp]: Doesn't this complain that there is no default? If not than, I am complaining :).

File: src/modules/vtol_att_control/tiltrotor.cpp [hunk 2]
```diff
@@ -218,47 +231,77 @@ void Tiltrotor::process_mc_data()
...
- }
 
-void Tiltrotor::process_fw_data()
-{
-	fill_att_control_output();
-}
+	_mc_roll_weight = 0.0f;
+	_mc_pitch_weight = 0.0f;
+	_mc_yaw_weight = 0.0f;
+ }
```
Comment (bansiesta) [inline src/modules/vtol_att_control/tiltrotor.cpp]: Whitespace. Actually, there is a couple of these.

File: src/modules/vtol_att_control/tiltrotor.cpp [hunk 3]
```diff
@@ -218,47 +231,77 @@ void Tiltrotor::process_mc_data()
...
-			_tilt_control = _params_tiltrotor.tilt_transition;
-		} else if (_tilt_control <= _params_tiltrotor.tilt_transition) {
-			_tilt_control = _params_tiltrotor.tilt_mc + fabsf(_params_tiltrotor.tilt_transition - _params_tiltrotor.tilt_mc) * 
-						(float) hrt_elapsed_time(&_vtol_schedule.transition_start) / (_params_tiltrotor.front_trans_dur * 1000000.0f);
+		if (_tilt_control <= _params_tiltrotor.tilt_transition) {
+			_tilt_control = _params_tiltrotor.tilt_mc +  fabsf(_params_tiltrotor.tilt_transition - _params_tiltrotor.tilt_mc)*(float)hrt_elapsed_time(&_vtol_schedule.transition_start)/(_params_tiltrotor.front_trans_dur*1000000.0f);
 		}
 
 		// do blending of mc and fw controls
-		if (_airspeed->true_airspeed_m_s >= ARSP_BLEND_START && _params_tiltrotor.airspeed_trans - ARSP_BLEND_START > 0.0f) {
```
Comment (bansiesta) [inline src/modules/vtol_att_control/tiltrotor.cpp]: Why did you remove this check?Are you sure that airspeed_trans > ARSP_BLEND_START?

File: src/modules/vtol_att_control/tiltrotor.cpp [hunk 4]
```diff
@@ -218,47 +231,77 @@ void Tiltrotor::process_mc_data()
...
+		_mc_yaw_weight = 1.0f;
+		if (_airspeed->true_airspeed_m_s > ARSP_YAW_CTRL_DISABLE) {
+			_mc_yaw_weight = 0.0f;
+		}
 
 	} else if (_vtol_schedule.flight_mode == TRANSITION_FRONT_P2) {
-		_tilt_control = _params_tiltrotor.tilt_transition +  fabsf(_params_tiltrotor.tilt_fw - _params_tiltrotor.tilt_transition) *
-					(float) hrt_elapsed_time(&_vtol_schedule.transition_start) / (0.5f * 1000000.0f);
-		_roll_weight_mc = 0.0f;
+		_tilt_control = _params_tiltrotor.tilt_transition +  fabsf(_params_tiltrotor.tilt_fw - _params_tiltrotor.tilt_transition)*(float)hrt_elapsed_time(&_vtol_schedule.transition_start)/(0.5f*1000000.0f);
```
Comment (bansiesta) [inline src/modules/vtol_att_control/tiltrotor.cpp]: This line is a bit long, and a comment would be nice.

File: src/modules/vtol_att_control/tiltrotor.cpp [hunk 5]
```diff
@@ -218,47 +231,77 @@ void Tiltrotor::process_mc_data()
...
+			set_rear_motor_state(ENABLED);
+		}
 		// tilt rotors forward up to certain angle
-		if (_params_tiltrotor.front_trans_dur <= 0.0f) {
-			_tilt_control = _params_tiltrotor.tilt_transition;
-		} else if (_tilt_control <= _params_tiltrotor.tilt_transition) {
-			_tilt_control = _params_tiltrotor.tilt_mc + fabsf(_params_tiltrotor.tilt_transition - _params_tiltrotor.tilt_mc) * 
-						(float) hrt_elapsed_time(&_vtol_schedule.transition_start) / (_params_tiltrotor.front_trans_dur * 1000000.0f);
+		if (_tilt_control <= _params_tiltrotor.tilt_transition) {
+			_tilt_control = _params_tiltrotor.tilt_mc +  fabsf(_params_tiltrotor.tilt_transition - _params_tiltrotor.tilt_mc)*(float)hrt_elapsed_time(&_vtol_schedule.transition_start)/(_params_tiltrotor.front_trans_dur*1000000.0f);
```
Comment (bansiesta) [inline src/modules/vtol_att_control/tiltrotor.cpp]: Did you check that _params_tiltrotor.front_trans_dur > 0?

Comment (bansiesta) [inline src/modules/vtol_att_control/tiltrotor.cpp]: Also way too long...

File: src/modules/vtol_att_control/tiltrotor.cpp [hunk 6]
```diff
@@ -218,47 +231,77 @@ void Tiltrotor::process_mc_data()
...
+	_v_att_sp->yaw_body = _yaw_transition;
+	_v_att_sp->thrust = _throttle_transition;
+
+	math::Matrix<3,3> R_sp;
+	R_sp.from_euler(_v_att_sp->roll_body,_v_att_sp->pitch_body,_v_att_sp->yaw_body);
+	memcpy(&_v_att_sp->R_body[0], R_sp.data, sizeof(_v_att_sp->R_body));
+
+	math::Quaternion q_sp;
+	q_sp.from_dcm(R_sp);
+	memcpy(&_v_att_sp->q_d[0], &q_sp.data[0], sizeof(_v_att_sp->q_d));
```
Comment (bansiesta) [inline src/modules/vtol_att_control/tiltrotor.cpp]: Shoudn't you set the valid flags here?

File: src/modules/vtol_att_control/tiltrotor.cpp [hunk 7]
```diff
@@ -310,7 +365,7 @@ Tiltrotor::set_max_fw(unsigned pwm_value)
 		if (i == 2 || i == 3) {
```
Comment (bansiesta) [inline src/modules/vtol_att_control/tiltrotor.cpp]: What does this mean? Please comment...

**Comment by bansiesta** [comment]
The cleanup/refactor looks good to me, apart from some small comments I made. I only had a brief look over the code but didn't go through the logic.

Note that your last 3 commits are somewhat not connected to a github account, probably check your git config.

**Comment by tumbili** [comment]
@bansiesta Thanks for the comments, really appreciate the review! Will have a look at them and do some fixes.

File: src/modules/vtol_att_control/tiltrotor.cpp (hunk 7 above)
Comment (LorenzMeier) [inline src/modules/vtol_att_control/tiltrotor.cpp]: This is missing a parameter to select which motors should be switched on / off.

**Comment by SimonWilks** [comment]
@tumbili can you rebase this on master now that vtol_cleanup is in? It will then be easier to review this.

**Comment by tumbili** [comment]
@SimonWilks Yes, will do!

**Comment by tumbili** [comment]
@SimonWilks Currently doing some flight tests, will do this later when I'm in the lab.

**Comment by SimonWilks** [comment]
@tumbili cool :)

**Comment by tumbili** [comment]
@SimonWilks Replaced by #2717
